File: continuum/reasoning/semantic_bridge.py
"""
Lightweight Semantic Causal Bridge (轻量语义因果桥接器).

Solves the fundamental "symptom-cause vocabulary gap" where symptoms describe
failure observations (e.g. 'SSLV3_ALERT_HANDSHAKE_FAILURE') while root causes describe
historical actions (e.g. 'updated openssl.conf with CipherString=DEFAULT@SECLEVEL=1').

Operates via Dual-Channel Query Projection:
  q_bridged = (1 - lambda) * q_symptom + lambda * q_hypotheses
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Callable, Sequence

import torch
from torch import Tensor


@dataclass(frozen=True)
class DiagnosticExpansion:
    """Result of diagnostic causal expansion."""
    raw_query: str
    matched_domains: list[str]
    expanded_concepts: list[str]
    expansion_text: str
    explanation: str


class SemanticCausalBridge:
    """
    Lightweight Diagnostic Causal Expansion Engine.
    Maps observable symptoms to candidate root-cause mechanism spaces in < 1 ms.
    """
    def __init__(self) -> None:
        self._rules: list[tuple[re.Pattern, list[str], str]] = []
        self._init_default_causal_domains()

    def _init_default_causal_domains(self) -> None:
        """Register default engineering and systems causal mappings."""
        # 1. TLS / SSL / Crypto Handshake Failures
        self.register_pattern(
            r"(?i)(sslv3|tls|handshake|handshake_failure|cipher|certificate|ssl_error|bad_mac)",
            [
                "openssl",
                "crypto",
                "cipherstring",
                "seclevel",
                "legacy_crypto",
                "openssl_conf",
                "ssl_version",
                "certificate_authority",
                "mutual_tls",
            ],
            domain="tls_crypto_configuration",
        )

        # 2. Database Connection Pool Starvation & 504 Timeouts
        self.register_pattern(
            r"(?i)(504\s+gateway\s+timeout|connection\s+pool|pool\s+exhausted|active_leases|lease_timeout)",
            [
                "db_connection_pool_size",
                "idle_timeout",
                "max_connections",
                "database_pool",
                "auth_service_config",
                "postgres_pool",
                "redis_pool",
            ],
            domain="database_resource_limits",
        )

        # 3. OOM / Memory Pressure / Garbage Collection Thrashing
        self.register_pattern(
            r"(?i)(oomkilled|out\s+of\s+memory|memoryerror|sigkill|rss\s+exceeded)",
            [
                "memory_limit",
                "heap_size",
                "buffer_allocation",
                "memory_leak",
                "gc_pause",
                "cgroup_memory",
                "resource_quota",
            ],
            domain="memory_limits_and_leaks",
        )

        # 4. Networking & DNS Resolution Failures
        self.register_pattern(
            r"(?i)(econnrefused|connection\s+refused|broken\s+pipe|dns_resolution|host\s+not\s+found|502\s+bad\s+gateway)",
            [
                "port_binding",
                "iptables",
                "upstream_proxy",
                "service_discovery",
                "coredns",
                "routing_table",
                "ingress_config",
            ],
            domain="network_ingress_routing",
        )

        # 5. Permission & Authentication Invalidation
        self.register_pattern(
            r"(?i)(permission\s+denied|eacces|403\s+forbidden|unauthorized|jwt\s+expired)",
            [
                "chmod",
                "file_mode",
                "rbac_binding",
                "iam_policy",
                "api_key_rotation",
                "service_account",
                "token_refresh",
            ],
            domain="security_rbac_credentials",
        )

        # 6. Dietary, Allergy & Life Safety Constraints
        self.register_pattern(
            r"(?i)(dietary|allergy|allergic|restriction|restrictions|intolerance|safety\s+restrictions)",
            [
                "peanut",
                "tree_nut",
                "nuts",
                "allergy",
                "severe_allergy",
                "dietary_safety",
                "lethal_allergy",
            ],
            domain="dietary_health_constraints",
        )

    def register_pattern(self, pattern: str, concepts: list[str], domain: str) -> None:
        """Register a new causal association rule."""
        compiled = re.compile(pattern)
        self._rules.append((compiled, concepts, domain))

    def expand(self, query_text: str) -> DiagnosticExpansion:
        """
        Analyze symptom text and extract hypothesized causal mechanism concepts.
        """
        matched_domains: list[str] = []
        expanded_concepts: list[str] = []

        for pattern, concepts, domain in self._rules:
            if pattern.search(query_text):
                matched_domains.append(domain)
                for c in concepts:
                    if c not in expanded_concepts:
                        expanded_concepts.append(c)

        if not expanded_concepts:
            # Fallback: extract technical keywords
            words = re.findall(r"\b[A-Za-z0-9_-]{4,}\b", query_text.lower())
            expanded_concepts = words[:8]
            matched_domains.append("generic_technical_keywords")

        expansion_text = " ".join(expanded_concepts)
        explanation = (
            f"Expanded {len(expanded_concepts)} causal hypotheses across {len(matched_domains)} "
            f"domains: {', '.join(matched_domains)}"
        )

        return DiagnosticExpansion(
            raw_query=query_text,
            matched_domains=matched_domains,
            expanded_concepts=expanded_concepts,
            expansion_text=expansion_text,
            explanation=explanation,
        )

    def project_query(
        self,
        query_text: str,
        embedder: Any,
        lambda_weight: float = 0.50,
    ) -> tuple[Tensor, DiagnosticExpansion]:
        """
        Produce a dual-channel bridged query vector.
        q_bridged = (1 - lambda) * q_symptom + lambda * q_hypothesis
        """
        expansion = self.expand(query_text)

        v_symptom = embedder.embed(query_text).detach().float()
        v_hypothesis = embedder.embed(expansion.expansion_text).detach().float()

        # Normalize components
        norm_s = torch.norm(v_symptom, p=2)
        if norm_s > 1e-8:
            v_symptom = v_symptom / norm_s

        norm_h = torch.norm(v_hypothesis, p=2)
        if norm_h > 1e-8:
            v_hypothesis = v_hypothesis / norm_h

        # Dual-channel linear fusion
        v_bridged = (1.0 - lambda_weight) * v_symptom + lambda_weight * v_hypothesis
        norm_b = torch.norm(v_bridged, p=2)
        if norm_b > 1e-8:
            v_bridged = v_bridged / norm_b

        return v_bridged, expansion

File: crates/continuum-core/src/semantic_bridge.rs
//! Native Semantic Causal Bridge in pure Rust.
//!
//! Bridges the symptom-cause vocabulary gap across domains (e.g. TLS handshake errors
//! to OpenSSL configuration changes) using dual-channel query projection in < 1 ms.

use crate::embedder::RealTextEmbedder;
use crate::math::normalize;

#[derive(Debug, Clone)]
pub struct DiagnosticExpansion {
    pub raw_query: String,
    pub matched_domains: Vec<String>,
    pub expanded_concepts: Vec<String>,
    pub expansion_text: String,
}

#[derive(Debug, Clone)]
pub struct SemanticCausalBridge {
    rules: Vec<(&'static [&'static str], &'static [&'static str], &'static str)>,
}

impl Default for SemanticCausalBridge {
    fn default() -> Self {
        Self::new()
    }
}

impl SemanticCausalBridge {
    pub fn new() -> Self {
        let rules: Vec<(&'static [&'static str], &'static [&'static str], &'static str)> = vec![
            // 1. TLS / SSL / Crypto Handshake Failures
            (
                &["sslv3", "tls", "handshake", "cipher", "certificate", "ssl_error", "bad_mac"],
                &[
                    "openssl",
                    "crypto",
                    "cipherstring",
                    "seclevel",
                    "legacy_crypto",
                    "openssl_conf",
                    "ssl_version",
                    "certificate_authority",
                    "mutual_tls",
                ],
                "tls_crypto_configuration",
            ),
            // 2. Database Connection Pool Starvation & 504 Timeouts
            (
                &["504", "gateway timeout", "connection pool", "pool exhausted", "active_leases"],
                &[
                    "db_connection_pool_size",
                    "idle_timeout",
                    "max_connections",
                    "database_pool",
                    "auth_service_config",
                    "postgres_pool",
                ],
                "database_resource_limits",
            ),
            // 3. Memory Pressure / OOM
            (
                &["oomkilled", "out of memory", "memoryerror", "sigkill", "rss exceeded"],
                &[
                    "memory_limit",
                    "heap_size",
                    "buffer_allocation",
                    "memory_leak",
                    "gc_pause",
                    "cgroup_memory",
                ],
                "memory_limits_and_leaks",
            ),
            // 4. Network Ingress & DNS Failures
            (
                &["econnrefused", "connection refused", "broken pipe", "dns", "502"],
                &[
                    "port_binding",
                    "iptables",
                    "upstream_proxy",
                    "service_discovery",
                    "coredns",
                    "ingress_config",
                ],
                "network_ingress_routing",
            ),
            // 5. Dietary, Allergy & Life Safety Constraints
            (
                &["dietary", "allergy", "allergic", "restriction", "restrictions", "intolerance", "safety restrictions"],
                &[
                    "peanut",
                    "tree_nut",
                    "nuts",
                    "allergy",
                    "severe_allergy",
                    "dietary_safety",
                    "lethal_allergy",
                ],
                "dietary_health_constraints",
            ),
        ];

        Self { rules }
    }

    /// Expands observable symptom text into candidate causal mechanism concepts.
    pub fn expand(&self, query_text: &str) -> DiagnosticExpansion {
        let lower = query_text.to_lowercase();
        let mut matched_domains = Vec::new();
        let mut expanded_concepts = Vec::new();

        for &(kws, concepts, domain) in &self.rules {
            let matched = kws.iter().any(|&kw| lower.contains(kw));
            if matched {
                matched_domains.push(domain.to_string());
                for &c in concepts {
                    let s = c.to_string();
                    if !expanded_concepts.contains(&s) {
                        expanded_concepts.push(s);
                    }
                }
            }
        }

        if expanded_concepts.is_empty() {
            matched_domains.push("generic_keywords".to_string());
        }

        let expansion_text = expanded_concepts.join(" ");

        DiagnosticExpansion {
            raw_query: query_text.to_string(),
            matched_domains,
            expanded_concepts,
            expansion_text,
        }
    }

    /// Projects query into dual-channel space:
    /// q_bridged = (1 - lambda) * q_symptom + lambda * q_hypothesis
    pub fn project_query(
        &self,
        query_text: &str,
        embedder: &RealTextEmbedder,
        lambda_weight: f32,
    ) -> (Vec<f32>, DiagnosticExpansion) {
        let expansion = self.expand(query_text);

        let v_symptom = embedder.embed(query_text);
        let v_hypothesis = embedder.embed(&expansion.expansion_text);

        let mut fused = vec![0.0f32; embedder.dim];
        for d in 0..embedder.dim {
            fused[d] = (1.0 - lambda_weight) * v_symptom[d] + lambda_weight * v_hypothesis[d];
        }

        (normalize(&fused), expansion)
    }
}
